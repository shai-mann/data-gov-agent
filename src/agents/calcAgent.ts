import { MessagesAnnotation, StateGraph } from '@langchain/langgraph';
import { ToolNode } from 'langchain';
import { AIMessage } from '@langchain/core/messages';
import { add } from '../tools';
import { openai } from '../llms';

// Set up tool calling
const tools = [add];
const llmWithTools = openai.bindTools(tools);

// Nodes
async function llmCall(state: typeof MessagesAnnotation.State) {
  // LLM decides whether to call a tool or not
  const result = await llmWithTools.invoke([
    {
      role: 'system',
      content:
        'You are a helpful assistant tasked with performing arithmetic on a set of inputs.',
    },
    ...state.messages,
  ]);

  return {
    messages: [result],
  };
}

const toolNode = new ToolNode(tools);

// Conditional edge function to route to the tool node or end
function shouldContinue(state: typeof MessagesAnnotation.State) {
  const messages = state.messages;
  const lastMessage = messages.at(-1) as AIMessage;

  // If the LLM makes a tool call, then perform an action
  if (lastMessage.tool_calls && lastMessage.tool_calls.length > 0) {
    return 'toolNode';
  }

  // Otherwise, we stop (reply to the user)
  return '__end__';
}

// Build workflow
const calcAgent = new StateGraph(MessagesAnnotation)
  .addNode('llmCall', llmCall)
  .addNode('toolNode', toolNode)
  // Add edges to connect nodes
  .addEdge('__start__', 'llmCall')
  .addConditionalEdges('llmCall', shouldContinue, ['toolNode', '__end__'])
  .addEdge('toolNode', 'llmCall')
  .compile();

export default calcAgent;
