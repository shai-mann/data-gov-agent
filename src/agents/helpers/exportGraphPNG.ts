import { CompiledStateGraph } from '@langchain/langgraph';
import sharp from 'sharp';

export default async function exportGraphPNG(
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  graph: CompiledStateGraph<any, any, any, any, any, any, any>
) {
  const drawableGraph = await graph.getGraphAsync();
  const jpegImage = await drawableGraph.drawMermaidPng(); // this is a lie - it is actually a JPEG image
  const jpegBuffer = await jpegImage.arrayBuffer();
  const pngBuffer = await sharp(jpegBuffer).png().toBuffer();
  return pngBuffer;
}
