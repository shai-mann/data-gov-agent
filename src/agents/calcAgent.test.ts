import { describe, beforeAll, afterAll, it, expect } from 'vitest';
import { setupNock } from '../../test/nockSetup';
import calcAgent from './calcAgent';
import redactInvocationIds from '../../test/redactInvocationIds';
import exportGraphPNG from './helpers/exportGraphPNG';

describe('Calc agent with Nock replay', () => {
  let nockHelper: ReturnType<typeof setupNock>;

  beforeAll(() => {
    nockHelper = setupNock('calc-agent');
  });

  afterAll(() => {
    nockHelper?.stop();
  });

  it('returns a deterministic response', async () => {
    const result = await calcAgent.invoke({
      messages: [
        {
          role: 'user',
          content: 'What is 2 + 2?',
        },
      ],
    });

    expect(redactInvocationIds(result)).toMatchSnapshot();
  });

  it('stores an agent graph', async () => {
    const pngBuffer = await exportGraphPNG(calcAgent);
    expect(pngBuffer).toMatchImageSnapshot();
  });
});
