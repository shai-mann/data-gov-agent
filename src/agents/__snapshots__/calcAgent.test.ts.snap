// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Calc agent with Nock replay > returns a deterministic response 1`] = `
{
  "messages": [
    {
      "additional_kwargs": {},
      "content": "What is 2 + 2?",
      "id": "REDACTED",
      "lc_kwargs": {
        "additional_kwargs": {},
        "content": "What is 2 + 2?",
        "id": "REDACTED",
        "response_metadata": {},
      },
      "lc_namespace": [
        "langchain_core",
        "messages",
      ],
      "lc_serializable": true,
      "name": undefined,
      "response_metadata": {},
      "type": "human",
      Symbol(langchain.message): true,
    },
    {
      "additional_kwargs": {
        "function_call": undefined,
        "tool_calls": [
          {
            "function": {
              "arguments": "{"a":2,"b":2}",
              "name": "add",
            },
            "id": "call_REDACTED",
            "type": "function",
          },
        ],
      },
      "content": "",
      "id": "chatcmpl-REDACTED",
      "invalid_tool_calls": [],
      "lc_kwargs": {
        "additional_kwargs": {
          "function_call": undefined,
          "tool_calls": [
            {
              "function": {
                "arguments": "{"a":2,"b":2}",
                "name": "add",
              },
              "id": "call_REDACTED",
              "type": "function",
            },
          ],
        },
        "content": "",
        "id": "chatcmpl-REDACTED",
        "invalid_tool_calls": [],
        "name": undefined,
        "response_metadata": {
          "model_name": "gpt-4o-mini-2024-07-18",
          "model_provider": "openai",
          "system_fingerprint": "fp_51db84afab",
          "usage": {
            "completion_tokens": 17,
            "completion_tokens_details": {
              "accepted_prediction_tokens": 0,
              "audio_tokens": 0,
              "reasoning_tokens": 0,
              "rejected_prediction_tokens": 0,
            },
            "prompt_tokens": 74,
            "prompt_tokens_details": {
              "audio_tokens": 0,
              "cached_tokens": 0,
            },
            "total_tokens": 91,
          },
        },
        "tool_calls": [
          {
            "args": {
              "a": 2,
              "b": 2,
            },
            "id": "call_REDACTED",
            "name": "add",
            "type": "tool_call",
          },
        ],
        "type": "ai",
        "usage_metadata": {
          "input_token_details": {
            "audio": 0,
            "cache_read": 0,
          },
          "input_tokens": 74,
          "output_token_details": {
            "audio": 0,
            "reasoning": 0,
          },
          "output_tokens": 17,
          "total_tokens": 91,
        },
      },
      "lc_namespace": [
        "langchain_core",
        "messages",
      ],
      "lc_serializable": true,
      "name": undefined,
      "response_metadata": {
        "finish_reason": "tool_calls",
        "model_name": "gpt-4o-mini-2024-07-18",
        "model_provider": "openai",
        "system_fingerprint": "fp_51db84afab",
        "tokenUsage": {
          "completionTokens": 17,
          "promptTokens": 74,
          "totalTokens": 91,
        },
        "usage": {
          "completion_tokens": 17,
          "completion_tokens_details": {
            "accepted_prediction_tokens": 0,
            "audio_tokens": 0,
            "reasoning_tokens": 0,
            "rejected_prediction_tokens": 0,
          },
          "prompt_tokens": 74,
          "prompt_tokens_details": {
            "audio_tokens": 0,
            "cached_tokens": 0,
          },
          "total_tokens": 91,
        },
      },
      "tool_calls": [
        {
          "args": {
            "a": 2,
            "b": 2,
          },
          "id": "call_REDACTED",
          "name": "add",
          "type": "tool_call",
        },
      ],
      "type": "ai",
      "usage_metadata": {
        "input_token_details": {
          "audio": 0,
          "cache_read": 0,
        },
        "input_tokens": 74,
        "output_token_details": {
          "audio": 0,
          "reasoning": 0,
        },
        "output_tokens": 17,
        "total_tokens": 91,
      },
      Symbol(langchain.message): true,
    },
    {
      "additional_kwargs": {},
      "artifact": undefined,
      "content": "4",
      "id": "REDACTED",
      "lc_direct_tool_output": true,
      "lc_kwargs": {
        "additional_kwargs": {},
        "artifact": undefined,
        "content": "4",
        "id": "REDACTED",
        "name": "add",
        "response_metadata": {},
        "tool_call_id": "call_REDACTED",
      },
      "lc_namespace": [
        "langchain_core",
        "messages",
      ],
      "lc_serializable": true,
      "name": "add",
      "response_metadata": {},
      "status": undefined,
      "tool_call_id": "call_REDACTED",
      "type": "tool",
      Symbol(langchain.message): true,
    },
    {
      "additional_kwargs": {
        "function_call": undefined,
        "tool_calls": undefined,
      },
      "content": "2 + 2 equals 4.",
      "id": "chatcmpl-REDACTED",
      "invalid_tool_calls": [],
      "lc_kwargs": {
        "additional_kwargs": {
          "function_call": undefined,
          "tool_calls": undefined,
        },
        "content": "2 + 2 equals 4.",
        "id": "chatcmpl-REDACTED",
        "invalid_tool_calls": [],
        "name": undefined,
        "response_metadata": {
          "model_name": "gpt-4o-mini-2024-07-18",
          "model_provider": "openai",
          "system_fingerprint": "fp_51db84afab",
          "usage": {
            "completion_tokens": 9,
            "completion_tokens_details": {
              "accepted_prediction_tokens": 0,
              "audio_tokens": 0,
              "reasoning_tokens": 0,
              "rejected_prediction_tokens": 0,
            },
            "prompt_tokens": 99,
            "prompt_tokens_details": {
              "audio_tokens": 0,
              "cached_tokens": 0,
            },
            "total_tokens": 108,
          },
        },
        "tool_calls": [],
        "type": "ai",
        "usage_metadata": {
          "input_token_details": {
            "audio": 0,
            "cache_read": 0,
          },
          "input_tokens": 99,
          "output_token_details": {
            "audio": 0,
            "reasoning": 0,
          },
          "output_tokens": 9,
          "total_tokens": 108,
        },
      },
      "lc_namespace": [
        "langchain_core",
        "messages",
      ],
      "lc_serializable": true,
      "name": undefined,
      "response_metadata": {
        "finish_reason": "stop",
        "model_name": "gpt-4o-mini-2024-07-18",
        "model_provider": "openai",
        "system_fingerprint": "fp_51db84afab",
        "tokenUsage": {
          "completionTokens": 9,
          "promptTokens": 99,
          "totalTokens": 108,
        },
        "usage": {
          "completion_tokens": 9,
          "completion_tokens_details": {
            "accepted_prediction_tokens": 0,
            "audio_tokens": 0,
            "reasoning_tokens": 0,
            "rejected_prediction_tokens": 0,
          },
          "prompt_tokens": 99,
          "prompt_tokens_details": {
            "audio_tokens": 0,
            "cached_tokens": 0,
          },
          "total_tokens": 108,
        },
      },
      "tool_calls": [],
      "type": "ai",
      "usage_metadata": {
        "input_token_details": {
          "audio": 0,
          "cache_read": 0,
        },
        "input_tokens": 99,
        "output_token_details": {
          "audio": 0,
          "reasoning": 0,
        },
        "output_tokens": 9,
        "total_tokens": 108,
      },
      Symbol(langchain.message): true,
    },
  ],
}
`;
